### PR TITLE
feat(forms): introduce InputGroup component

### PR DESCRIPTION
**Pull request by austin94**
## Description

This PR introduces the `InputGroup` component. It allows Buttons (and other elements) to be prepended, appended around an `Input` element.

Screen Shot 2020-08-18 at 11 19 55 AM

```tsx
<Field>
  <Label>Input group</Label>
  <InputGroup>
    <Button>A</Button>
    <Input />
    <Button>B</Button>
  </InputGroup>
</Field>
```

### `focusInset` styling

To ensure consistent focus styling I'm using a context object to inform the `Input` element to default to use `focusInset` styling. Since the `Prepend/Append` elements can receive any element it is up to consumers to apply the props to the element if they're using a Garden `Button`.

### Field sizing

The `InputGroup` element disrupts the current styling between labels, hints, messages, and input. To ensure consistent styling I've included an `isCompact` prop to `InputGroup` which is also passed to the `Input` through context.

## Checklist

- [x] :ok_hand: design updates are Garden Designer approved (add the
      designer as a reviewer)
- [x] :globe_with_meridians: Styleguidist demo is up-to-date (`yarn start`)
- [x] :arrow_left: renders as expected with reversed (RTL) direction
- [x] :metal: renders as expected with Bedrock CSS (`?bedrock`)
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :guardsman: includes new unit tests
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

### CODE REVIEW

**Review by m-lai — APPROVED** [review]
Looks good to me. I think if we allow multiple buttons/elements then `Prepend` and `Append` make sense. Sketch symbol naming will still follow the Split Button naming format I reckon, so it's easier for designers to use:

1. input-group-default/1. primary/1. default/1. default
1. input-group-default/1. primary/1. default/2a. hover-input
1. input-group-default/1. primary/1. default/2b. hover-button
1. input-group-default/1. primary/1. default/3. active-button
1. input-group-default/1. primary/1. default/4a. focus-input

**Comment by austin94** [comment]
@jzempel weird, this is only showing on iOS mobile.

Screen Shot 2020-08-20 at 1 42 14 PM

Looking better to me know.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import { StyledMessage } from '../common/StyledMessage';
+
+const COMPONENT_ID = 'forms.input_group';
+
+interface IStyledInputGroupProps {
+  isCompact?: boolean;
+}
+
+const positionStyles = (props: ThemeProps<DefaultTheme> & IStyledInputGroupProps) => {
+  const topMargin = math(`${props.theme.space.base} * ${props.isCompact ? '1px' : '2px'}`);
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledInputGroup.ts:23]: Updated to remove the `math` utility.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      /* stylelint-enable */
+    }
+  `;
+};
+
+export const StyledInputGroup = styled.div.attrs({
+  'data-garden-id': COMPONENT_ID,
+  'data-garden-version': PACKAGE_VERSION
+})<IStyledInputGroupProps>`
+  display: flex;
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledInputGroup.ts:67]: > Is there an advantage for making this occupy the entire block?

Our inputs are `display: block` by default I thought. Would we want these groups to be inline?

> should direction be getting applied based on RTL?

The parent `Field` already applies that for us!

File: packages/forms/src/styled/input-group/StyledInputGroup.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      margin-top: ${topMargin};
+    }
+    /* stylelint-enable */
+
+    ${StyledTextInput} {
+      position: relative;
+      flex: 1 1 auto;
+      margin-top: 0;
+      margin-bottom: 0;
+      width: 1px;
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledInputGroup.ts:43]: This overrides the default `width: 100%` of the `StyledTextInput`

Screen Shot 2020-08-24 at 2 37 49 PM

I've added an inline comment for this.

File: packages/forms/examples/input-group.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    <InputGroup.Prepend>
+      <Button focusInset>B</Button>
+    </InputGroup.Prepend>
+    <Input placeholder="Input content" />
+    <InputGroup.Append>
+      <Button focusInset isPrimary>
+        Copy
+      </Button>
+    </InputGroup.Append>
+  </InputGroup>
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/forms/examples/input-group.md:26]: This is handled correctly I believe. The styling is applied when a sibling of the `StyledLabel`

```
${StyledLabel} + & {
  /** margin-top styles */
}
```

File: packages/forms/examples/input-group.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  - [`isCompact`] Apply compact styling
+- `InputGroup.Prepend` Accepts all `div` attributes
+- `InputGroup.Append` Accepts all `div` attributes
+
+```jsx
+const { Button } = require('@zendeskgarden/react-buttons/src');
+
+<Field>
+  <Label>Input Group</Label>
+  <InputGroup>
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/forms/examples/input-group.md:13]: I've added a knobs section with an `isCompact` toggle

File: packages/forms/examples/input-group.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (austin94) [reply packages/forms/examples/input-group.md:26]: I've added a validation message to the new knobs example to show this.

File: packages/forms/src/styled/input-group/StyledPrepend.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    /* stylelint-enable */
+  `;
+};
+
+export const StyledPrepend = styled.div.attrs({
+  'data-garden-id': COMPONENT_ID,
+  'data-garden-version': PACKAGE_VERSION
+})`
+  display: flex;
+  z-index: 2;
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledPrepend.ts:39]: This was tricky to bring over, but I think I got it working. We need to start out the `z-index` at `1` with the `InputGroup` to beat the `StyledTextInput` styling.

I've added similar z-index logic to the `Prepend` and `Append` components.

image

File: packages/forms/src/styled/input-group/StyledAppend.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  z-index: 1;
+
+  &:hover,
+  &:active {
+    z-index: 2;
+  }
+
+  &:disabled {
+    z-index: -1;
+  }
```
Comment (jzempel) [inline packages/forms/src/styled/input-group/StyledAppend.ts:48]: Disabled styling doesn't appear to be playing as expected. This div cannot be disabled. But when it's child button is disabled, the hover styling gets wonky. Styling should reflect, to some degree, the disabled styling seen with https://zendeskgarden.github.io/react-components/buttons/#!/Basic/9.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+      margin-top: ${topMargin};
+    }
+    /* stylelint-enable */
+
+    ${StyledTextInput} {
+      position: relative;
+      flex: 1 1 auto;
+      margin-top: 0;
+      margin-bottom: 0;
+      width: 1px; /* [1] */
```
Comment (jzempel) [inline packages/forms/src/styled/input-group/StyledInputGroup.ts:45]: I think you might want `width: auto;` here to override and reset.

File: packages/forms/src/styled/input-group/StyledAppend.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+export const StyledAppend = styled.div.attrs({
+  'data-garden-id': COMPONENT_ID,
+  'data-garden-version': PACKAGE_VERSION
+})`
+  display: flex;
+  z-index: 1;
+
+  &:hover,
+  &:active {
+    z-index: 2;
```
Comment (jzempel) [inline packages/forms/src/styled/input-group/StyledAppend.ts:43]: > This was tricky to bring over, but I think I got it working. We need to start out the z-index at 1 with the InputGroup to beat the StyledTextInput styling.

I'm not seeing that `StyledTextInput` has a `z-index`.

File: packages/forms/src/styled/input-group/StyledAppend.ts (hunk 7 above)
Comment (jzempel) [reply packages/forms/src/styled/input-group/StyledAppend.ts:48]: Could be that the example should be utilizing a `ButtonGroup` within the `InputGroup` in order to get the styling right?

File: packages/forms/src/styled/input-group/StyledInputGroup.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+      /* stylelint-enable */
+    }
+  `;
+};
+
+export const StyledInputGroup = styled.div.attrs({
+  'data-garden-id': COMPONENT_ID,
+  'data-garden-version': PACKAGE_VERSION
+})<IStyledInputGroupProps>`
+  display: inline-flex;
```
Comment (jzempel) [inline packages/forms/src/styled/input-group/StyledInputGroup.ts:69]: > Our inputs are display: block by default I thought.

Garden's text inputs are `display: inline-block`, making `inline-flex` the analog.

File: packages/forms/src/styled/input-group/StyledAppend.ts (hunk 7 above)
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledAppend.ts:48]: > disabled styling doesn't appear to be playing as expected

I was accidentally applying `z-index` to the wrapping div. This now targets the child elements correctly.

Comment (austin94) [reply packages/forms/src/styled/input-group/StyledAppend.ts:48]: I'm not sure using a `ButtonGroup` would work here. It includes semantic roles and keyboard interaction that diverges from this pattern.

File: packages/forms/src/styled/input-group/StyledAppend.ts (hunk 9 above)
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledAppend.ts:43]: Without starting at `z-index: 1` we have overlay issues with the input 

image

Comment (jzempel) [reply packages/forms/src/styled/input-group/StyledAppend.ts:43]: Got it. In that case, wouldn't `StyledTextInput` start at `z-index: -1` and then apply `z-index: 1` on hover/active?

**Comment by austin94** [comment]
@jzempel I've updated this to remove the `Prepend/Append` components as requested and added a `disabled` knob to the example.

I'm still experiencing some z-index weirdness with the text input due to the negative margin we apply. I've bumped the z-index to fix this, but want to double check with you.

File: packages/forms/src/styled/input-group/StyledAppend.ts (hunk 9 above)
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledAppend.ts]: `z-index: -1` begins to place elements behind the body at that point

image

Comment (jzempel) [reply packages/forms/src/styled/input-group/StyledAppend.ts]: This is why resetting the stacking context by applying `z-index: 0` to the `InputGroup` wrapper is so important. Once that is done, the children only need to receive either `z-index: 1` or `z-index: -1`.

Comment (jzempel) [reply packages/forms/src/styled/input-group/StyledAppend.ts]: See https://developer.mozilla.org/en-US/docs/Web/CSS/z-index#Values for details.

File: packages/forms/examples/input-group.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+                  <Code>isCompact</Code>
+                </Label>
+              </Toggle>
+            </Field>
+          </Col>
+          <Col>
+            <Field>
+              <Toggle
+                checked={state.isDisabled}
+                onChange={event => setState({ isDisabled: event.target.checked })}
```
Comment (jzempel) [inline packages/forms/examples/input-group.md:39]: The disabled state should also apply to the input field.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+      margin-${props.theme.rtl ? 'right' : 'left'}: -${props.theme.borderWidths.sm};
+    }
+
+    & > button {
+      z-index: 1;
+    }
+
+    & > button:hover,
+    & > button:active {
+      z-index: 2;
```
Comment (jzempel) [inline packages/forms/src/styled/input-group/StyledInputGroup.ts:78]: Once stacking context is applied to the `InputGroup`, this should be z-index: 1. Also the rule here should be for all of:
```css
& > ${StyledInput}:hover,
& > button:hover,
& > ${StyledInput}:active,
& > button:active {
  z-index: 1;
}
```

File: packages/forms/src/styled/input-group/StyledInputGroup.ts [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+const itemStyles = (props: ThemeProps<DefaultTheme>) => {
+  return css`
+    /* stylelint-disable property-no-unknown, property-case */
+    & > * {
+      margin-${props.theme.rtl ? 'right' : 'left'}: -${props.theme.borderWidths.sm};
+    }
+
+    & > button {
+      z-index: 1;
+    }
```
Comment (jzempel) [inline packages/forms/src/styled/input-group/StyledInputGroup.ts:74]: Once stacking context is applied to the `InputGroup`, replace this with:

```css
& > ${StyledInput} {
  z-index: -1;
}
```

File: packages/forms/src/styled/input-group/StyledInputGroup.ts [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+    & > button:hover,
+    & > button:active {
+      z-index: 2;
+    }
+
+    & > button:disabled {
+      z-index: 0;
```
Comment (jzempel) [inline packages/forms/src/styled/input-group/StyledInputGroup.ts:82]: Once stacking context is applied to the `InputGroup`, this should be `z-index: -1`.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+      border-top-width: 0;
+      border-bottom-width: 0;
+    }
+
```
Comment (jzempel) [inline packages/forms/src/styled/input-group/StyledInputGroup.ts:86]: Once stacking context is applied to `InputGroup`, add:

```css
& > ${StyledInput} {
  z-index: -2;
}
```

This is the one case that drops outside the [-1 : 1] threshold.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    & > button:last-of-type:not(:first-of-type) {
+      border-top-${props.theme.rtl ? 'right' : 'left'}-radius: 0;
+      border-bottom-${props.theme.rtl ? 'right' : 'left'}-radius: 0;
+    }
+
+    & > button:not(:first-of-type):not(:last-of-type) {
+      border-radius: 0;
+    }
```
Comment (jzempel) [inline packages/forms/src/styled/input-group/StyledInputGroup.ts:100]: These should not be qualified with `button`. Removing elements breaks styling.

File: packages/forms/examples/input-group.md (hunk 11 above)
Comment (austin94) [reply packages/forms/examples/input-group.md:39]: Updated.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts (hunk 12 above)
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledInputGroup.ts:78]: Stacking context has been updated.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts (hunk 13 above)
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledInputGroup.ts:74]: Stacking context has been updated.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts (hunk 14 above)
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledInputGroup.ts:82]: Stacking context has been updated.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts (hunk 15 above)
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledInputGroup.ts:86]: Stacking context has been updated.

File: packages/forms/src/styled/input-group/StyledInputGroup.ts (hunk 16 above)
Comment (austin94) [reply packages/forms/src/styled/input-group/StyledInputGroup.ts:100]: I've updated this to use `:first-child` and `:last-child` pseudo selectors.

**Comment by jzempel** [comment]
blerg, these false-positive deploy tests should get solved by the `@zendeskgarden/scripts` update in #845. In the meantime, I'll pull and test locally.

File: packages/forms/examples/input-group.md [hunk 17]
```diff
@@ -0,0 +1,83 @@
+### Components
+
+- `InputGroup` Accepts all `div` attributes and the following props:
+  - [`isCompact`] Apply compact styling
+- `InputGroup.Prepend` Accepts all `div` attributes
+- `InputGroup.Append` Accepts all `div` attributes
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/forms/examples/input-group.md]: Updated both.

[the remark this answers is not in the document]
Comment (jzempel) [reply packages/forms/examples/input-group.md]: [nit] the PR description is still referring to Append/Prepend.

[the remark this answers is not in the document]
Comment (austin94) [reply packages/forms/examples/input-group.md]: Missed the description shown in the image background. Image has been updated now.

**Pull request merged**